### PR TITLE
Implement groupby in pylibcudf

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR implements groupby in pylibcudf along with the minimal set of aggregation logic to support groupby. To limit its scope, this PR does not include other aggregation logic for e.g. non-groupby reductions and scans. Due to the large scale of what's already in this PR, I have also omitted the changes required to leverage pylibcudf in the current cudf Cython code from this PR. That will be done in a follow-up. This PR's diff is misleadingly large, a large chunk of it is adding documentation and function declarations that shouldn't impose too heavy a cognitive load in review.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/pylibcudf/groupby.pyx [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+cdef class GroupByRequest:
+    """A request for a groupby aggregation or scan.
+
+    Parameters
+    ----------
+    values : Column
+        The column to aggregate.
+    aggregations : list
+        The list of aggregations to perform.
```
Comment (shwina) [inline python/cudf/cudf/_lib/pylibcudf/groupby.pyx:36]: Maybe we can be more specific here (`List[Aggregation]`)?

File: python/cudf/cudf/_lib/pylibcudf/groupby.pyx [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    """
+    def __init__(self, Column values, list aggregations):
+        self.values = values
+        self.aggregations = aggregations
+
+    cdef aggregation_request to_libcudf_agg_request(self) except *:
+        """Convert to a libcudf aggregation_request object.
+
+        This method is for internal use only. It creates a new libcudf
```
Comment (shwina) [inline python/cudf/cudf/_lib/pylibcudf/groupby.pyx:45]: In general should these be prefixed with `_`?

File: python/cudf/cudf/_lib/pylibcudf/groupby.pyx [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        Table keys,
+        null_policy null_handling=null_policy.EXCLUDE,
+        sorted keys_are_sorted=sorted.NO
+    ):
+        self.c_obj.reset(new groupby(keys.view(), null_handling, keys_are_sorted))
+
+    @staticmethod
+    cdef tuple _parse_outputs(
+        pair[unique_ptr[table], vector[aggregation_result]] c_res
+    ):
```
Comment (shwina) [inline python/cudf/cudf/_lib/pylibcudf/groupby.pyx:92]: One liner describing what this function does would be helpful

**Review by shwina — APPROVED** [review]
Looks good to me - just some small doc requests.

File: python/cudf/cudf/_lib/pylibcudf/groupby.pyx (hunk 2 above)
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/groupby.pyx:45]: Yeah I think so.

File: python/cudf/cudf/_lib/pylibcudf/groupby.pyx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        # Convert libcudf aggregation/scan outputs into pylibcudf objects.
+        # This function is for internal use only.
+        cdef Table group_keys = Table.from_libcudf(move(c_res.first))
+
+        cdef int i, j
```
Comment (mroeschke) [inline python/cudf/cudf/_lib/pylibcudf/groupby.pyx:101]: In pandas we typically use `Py_ssize_t` for loop variables, but that shouldn't make too much of a difference compared to `int`

**Review by mroeschke — APPROVED** [review]

File: python/cudf/cudf/_lib/pylibcudf/groupby.pyx (hunk 4 above)
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/groupby.pyx:101]: Yeah that is probably technically more correct, but agree it shouldn't matter much in our Cython code.

**Pull request merged**